Refactor HTTP connection handling and some handlers to stream responses

jschmidt-icinga opened the pull request:
Fixes #10409.
Fixes #10142, since many of the additional parameters `HttpHandler`s needed before are no longer necessary.

# Description

This PR mostly does four things:

- Simplify the `HttpHandler` signature by containing additional parameters and interfaces in classes extending the `boost::beast` request and response classes.
- Add generic interface and connection handling for `HttpHandler`s to stream responses via chunked encoding.
- Refactor the `ObjectQueryHandler` and `EventsHandler` to make use of these streaming capabilities.
- Add unit-tests for `HttpServerConnection` and the new `HttpResponse` and `HttpRequest` classes.

I will explain the implementation and the considerations I've made in detail below.

## `HttpRequest` and `HttpResponse`

The rationale for adding these classes is that they serve as an abstraction between `HttpServerConnection`, which does no longer have to offer an interface to the handlers directly (i.e. via a `server` parameter passed down to the handlers and `StartStreaming()` being called by the handlers), and the `HttpHandlers`, which do no longer need to concern themselves with connection handling.

For the response class I've implemented a [custom body type](https://www.boost.org/doc/libs/latest/libs/beast/doc/html/beast/using_http/custom_body_types.html) `SerializableBody` with a [`writer`](https://www.boost.org/doc/libs/latest/libs/beast/doc/html/beast/concepts/BodyWriter.html) that can be used by the [serializer](https://www.boost.org/doc/libs/latest/libs/beast/doc/html/beast/ref/boost__beast__http__response_serializer.html) to write the added content to the connection, while adding chunked encoding as needed and tracking progress as needed. A previous iteration of this PR also implemented the [`BodyReader`](https://www.boost.org/doc/libs/latest/libs/beast/doc/html/beast/concepts/BodyReader.html), but this added unnecessary complexity and was removed.

This wasn't easily achievable with any of the regular beast body types, because we needed functionality of the `buffer_body`, specifically the ability to interrupt serialization when the buffer content has fully been written to the connection and pick up again on the next call to `Flush()` once more data is available, and the `dynamic_body`, which has the ability to allocate additional memory as needed.

### Streaming and non-streaming responses

To start a streaming response, the handler has to first call `response.StartStreaming()`, which enables chunked encoding and allows the response header to be sent on the next `response.Flush(yc)` without setting the `content_length`.

In case of a non-streaming response, the handler can use the `HttpUtility::SendJson(Body|Error)` convenience functions, same as before, or write the body as normal and then just return. The handler no longer has to set up `content_length`, which is now done automatically in the first call to `response.Flush(yc)` which by default is done once after the handler returns to `ProcessRequest()` in `httpserverconnection.cpp`.

Due to the new custom body, it is no longer possible (it would have been possible to implement, but more work) to use string operators on the body, however an output stream operator is added to make the interface as or more convenient. There is however a bug in beast that makes this slightly less efficient (but still more efficient than string `+=`) than it could be (see the comment in [here](https://github.com/Icinga/icinga2/blob/d5cedf17c1aaf9214f4068a55d55c2461d9ba1f9/lib/remote/httpmessage.hpp#L40-L58)). I intend to write a minimal reproducer and report this bug to the beast github repo.

### Refactoring `ObjectQueryHandler` and `EventsHandler`

Events were streamed before, but without using chunked encoding and needing to directly setup the server connection via `StartStreaming()`. Now, proper HTTP/1.1 compliant chunked encoding is used and no special treatment is necessary by the server connection.

In `ObjectQueryHandler` we now make use of the new ability of passing a generator function to the `JsonEncoder` added in #10414. The for-loop iterating over the config objects becomes a generator function that feeds the serialized objects into the JsonEncoder one by one, flushing the response occasionally.

The same could easily be applied to other handlers like `ModifyObjectHandler` and `DeleteObjectHandler`, but since these don't have as large of a memory footprint, the necessity of this can still be determined and done in a future PR.

## `HttpServerConnection::DetectClientShutdown`

This new coroutine was added to `HttpServerConnection` to detect shutdowns initiated by the HTTP-client while we're streaming a response. Before this PR, we'd detect shutdowns either when writes failed, or when the next request was being read. This had the disadvantage that during long running responses there were situations where it was impossible for the client to complete a graceful SSL-shutdown within a reasonable time-frame. Even with the new streaming capabilities added by this PR, waiting until writes failed would lead to an "application data after close notify" error on the client's side and an exception being logged on the server's side.

In my opinion adding a general approach as a part of `HttpServerConnection`'s connection handling made the most sense, and this coroutine was the simplest way to do this. Possibly there would have been ways to let handlers enable or disable this functionality, but that would have been more code for basically no benefit, as this coroutine doesn't do any harm in any of the other use-cases.

The Coroutine basically initiates an `async_fill()` whenever reading a request is done:

and for as long as the connection isn't in the process of (or done) disconnecting:

https://github.com/Icinga/icinga2/blob/d5cedf17c1aaf9214f4068a55d55c2461d9ba1f9/lib/remote/httpserverconnection.cpp#L532-L533

Initially I intended to use `boost::asio`'s [`async_wait`](https://www.boost.org/doc/libs/latest/doc/html/boost_asio/reference/basic_stream_socket/async_wait.html) function with `wait_type::wait_read` to avoid even as much as an `async_fill` in the case a client didn't initiate a shutdown. However, likely due to differences in the respective libc's, this would not work reliably on the Windows and Alpine builds. So instead I chose the following approach:

https://github.com/Icinga/icinga2/blob/d5cedf17c1aaf9214f4068a55d55c2461d9ba1f9/lib/remote/httpserverconnection.cpp#L535-L541

This tries to fill the stream's buffer with as little as a single byte from the connection, which either yields until one or more bytes are readable, or returns an error code when the connection is closed for whatever reason. In any case where an error code is returned, we will want to attempt a graceful shutdown, but mostly this will be responding to a shutdown the client has initiated.

In the case no error is returned and the connection is being reused for another request, the first few bytes of the request will be read into the stream's buffer and an `AsioDualEvent` flag will be set to allow the regular continuation of `HttpServerConnection::ProcessMessage()` to read the next request. This "locking" is needed to avoid two `async_fill`s (the one here and the one in `http::async_read`) being in progress at the same time, which causes exceptions/errors under certain race-conditions.

## Unit Testing

The largest part of new lines in this PR comes from the addition of unit testing for `HttpServerConnection` and the new HTTP message classes. To get this working I had to add a number of testing fixtures for setting up certificates, TLS connections, and log message pattern matching. These were all added in a generally useful way so its possible to reuse them for adding other medium complexity unit-tests later on. After this PR is merged, I intend to use these fixtures to add testing for `ApiListener`, `JsonRpcConnection`, and some individual `HttpHandler`s.

Generating the certificates is relatively time consuming and takes above 1.5s on my system, so we wouldn't want to do it once per test, which is why I've added cmake-test fixtures (not to be confused with boost-test fixtures) which are implemented as "test-cases" themselves that set up the certificates in a temporary directory and define a dependency to all tests that use them. Ctest will then schedule the setup fixture test to be run before and the cleanup fixture test to be run after all dependent tests are finished. This looks a bit verbose inside `CMakeLists.txt`, which stems from our `add_boost_test()` cmake function being old and limited legacy code. Maybe this can be revisited in the future to make it more elegant, like automated discovery of test targets similar to `doctest`'s cmake integration. Anyway, after the initial setup of the certificates, this makes each test case take about 80-100ms on my system, with the exception of `remote_httpserverconnection/liveness_disconnect` which has to wait 10s for the disconnect to be triggered in `HttpServerConnection::CheckLiveness()`.

For `HttpServerConnection` I've focused on testing connection handling with a simple test handler, so it doesn't rely on or test any specific behavior of the regular handlers. I've verified that all tests (including `client_shutdown` using `HttpServerConnection::StartStreaming` in the handler) also run on the master branch with some modifications to the`UnitTestHandler` and some of the Log Pattern matching commented out. I've decided against making them a separate PR, because then almost every commit in this PR would have to incrementally adapt that handler to the changes to still make it compile.

For `HttpResponse` and `HttpRequest`, parsing and serializing the messages is tested, in streaming and non-streaming conditions, and also their use together with some of the auxiliary functions in `HttpUtility`.

If anyone can think of test cases I missed, I'm be happy to add them.

---

## Tests (yhabteab)

On a Debian VM of the following size:
```bash
RAM - 8GB
VCPUs - 4 VCPU
Disk - 50GB
```

Icinga 2 is running as a container within that VM using a locally built image from this PR. So, there is no other process involved in the resulted graph, since I've generated them using `docker stats` with a timeframe of `30m` and Icinga 2 is the only container running.

I've used this script from outside of the VM (from my local machine) to stress test it:
```bash
#!/bin/zsh

# Usage:
#
#   while :; do ./do-request.sh 100; printf .; done

for ((i = 0; i < "$1"; i++)); do
    curl -sSk \
    		-u root:a8f80869d65de8ae \
    		-o /dev/null \
    		'https://10.27.0.95:5665/v1/objects/services?pretty=1' &
done

wait "$(jobs -p)"
```

<details><summary>icinga2 daemon -C</summary>

```bash
[2025-08-01 10:09:00 +0000] information/cli: Icinga application loader (version: v2.15.0-68-ge9f7f1c5f)
[2025-08-01 10:09:00 +0000] information/cli: Loading configuration file(s).
[2025-08-01 10:09:01 +0000] information/ConfigItem: Committing config item(s).
[2025-08-01 10:09:01 +0000] information/ApiListener: My API identity: icinga-master
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 NotificationComponent.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 CheckerComponent.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 16124 Dependencies.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 User.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 UserGroup.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 3 TimePeriods.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 3 ServiceGroups.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 12419 Services.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 ScheduledDowntime.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 3 Zones.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 12 Notifications.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 2 NotificationCommands.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 FileLogger.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 IcingaApplication.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 530 Hosts.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 2 HostGroups.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 Endpoint.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 ApiUser.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 1 ApiListener.
[2025-08-01 10:09:03 +0000] information/ConfigItem: Instantiated 248 CheckCommands.
```

</details> 

**Before:**

It couldn’t even run for straight **1 minute** before the OOM-Killer was triggered:
```bash
[14170.329916] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=containerd.service,mems_allowed=0,global_oom,task_memcg=/system.slice/docker-369ef8f43cd345cbd20e290bb24d8d209f2dda377ff4e9b77170a4b72fca7036.scope,task=icinga2,pid=106144,uid=5665
[14170.333017] Out of memory: Killed process 106144 (icinga2) total-vm:8940468kB, anon-rss:7622224kB, file-rss:204kB, shmem-rss:0kB, UID:5665 pgtables:15304kB oom_score_adj:0
```

<img width="1000" height="500" alt="CPU %" src="https://github.com/user-attachments/assets/64e7cf95-53dd-4e13-a6ff-c3dcd47fc1f1" />
<img width="1000" height="500" alt="MEM %" src="https://github.com/user-attachments/assets/5839a337-f97c-4a12-b80a-96685031d139" />

**After:**
<img width="1000" height="500" alt="CPU %" src="https://github.com/user-attachments/assets/46b26fba-24d2-49c5-81fc-8ee3ea189b75" />
<img width="1000" height="500" alt="MEM %" src="https://github.com/user-attachments/assets/a78bac17-ad13-4fab-af8a-c3d499e21ad7" />

jschmidt-icinga commented on lib/remote/httpmessage.cpp:
```diff
@@ -0,0 +1,167 @@
+HttpRequest::HttpRequest(Shared<AsioTlsStream>::Ptr stream)
+	: m_Stream(std::forward<decltype(m_Stream)>(stream)){}
+
+void HttpRequest::ParseHeader(boost::beast::flat_buffer & buf, boost::asio::yield_context yc)
```
Right. I've added it to my `.clang-format` file now, so this hopefully will not happen again (same for pointers).

julianbrost replied:
> I've added it to my `.clang-format` file now

If you've already put in the time and effort for creating one, would you mind sharing it, i.e. creating a PR for it? Even if it isn't perfect yet, I'd use it locally to see how well it works and suggest changes if I notice problems. Ideally, one nice day we can integrate it, have a CI check for it and remove the need for review comments like this one. :)

jschmidt-icinga commented on lib/remote/httpmessage.cpp:
```diff
@@ -0,0 +1,167 @@
+	body() = std::move(m_Parser.release().body());
+}
+
+ApiUser::Ptr HttpRequest::User() const
```
RVO should make sure there aren't any copies (i.e. increments to the ref-count) and while this is unlikely to happen here, returning references to members of objects with non-static lifetimes is often a bad idea. 

HttpRequests are unlikely to be temporary objects, but it's still good practice and as explained, we're not losing anything due to RVO/copy-elision.

yhabteab replied:
Indeed!

jschmidt-icinga commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -588,3 +519,25 @@ void HttpServerConnection::CheckLiveness(boost::asio::yield_context yc)
+		boost::system::error_code ec;
+		m_Stream->async_fill(yc[ec]);
+		if (ec && !m_ShuttingDown) {
+			Log(LogInformation, "HttpServerConnection") << "Detected shutdown from client: " << m_PeerAddress << ".";
```
The log entry is not identical and it's not there for unit-testing, it is tested for in unit-testing. I wanted a way to distinguish between disconnects we initiated on our side or the client initiated on their side. I added the relevant asserts to the tests to make sure this message is only triggered when it makes sense.

yhabteab replied:
> I wanted a way to distinguish between disconnects we initiated on our side or the client initiated on their side.

If it's **not** for unittesting, then in which situations would this distinction be relevant whether the shutdown is initiated from some random HTTP client or Icinga 2 itself? In most cases, the client will close the connection and you’ll end up with two identical<sup>at least to me</sup> log entries that basically says "client hung up me"  and "connection closed", but show no further differences. If that's just me, then fine you can leave it as-is, but it will further clutter the otherwise already crowded Icinga 2 logs.

jschmidt-icinga replied:
We have messages in the log that are far more useless than this one, and setting it to LogInformation is analogous to the one in `CheckLiveness()`, which IMHO would also benefit from printing the address btw., since you can't tell from the message which connection out of potentially dozens it is that timed out.

But I'd be fine with downgrading it to LogDebug or LogNotice. But it is valid debug information at the very least, if anything we should add even more traces to see what's going on with connection handling in case of bug reports, or to contextualize a time-line of events that led to some other bug. Especially disconnects currently can be triggered from several places with no idea how or where the disconnect happened.

yhabteab replied:
> We have messages in the log that are far more useless than this one, and setting it to LogInformation is analogous to the one in `CheckLiveness()`

Absolutely, but think of a typical Icinga 2 HTTP request: Normally, you send an API request and close the connection immediately after you get what you requested, meaning that in almost all cases it's the client that tears down the connection, and in all those cases the logs will be flooded with such entries. Actually, there’s nothing wrong in closing the connection that you don't (won't) need and so we shouldn't have to log every time - oh the client is gone - twice. The one in `CheckLiveness()`, on the other hand, is not to be expected, and if it does happen, it is a sign that something is wrong either with Icinga2 or on the client side. So it is worth logging this, but I agree that it would also be beneficial to add more information about the client.

> But I'd be fine with downgrading it to LogDebug or LogNotice.

As I said in my previous comment: If I am the only one who does not yet see its value, then you don't have to remove it completely or demote it to debug or notice. I am just trying to explain what I am currently observing when interacting with the Icinga 2 API.

jschmidt-icinga commented:
- Added `HttpResponse::SendFile()` method (WIP)
- Reformatted new code with `.clang-format` (please tell me if you find anything that doesn't match the style convention)
- Fixed potential concurrency issues with Boost.Test asserts

jschmidt-icinga commented on lib/remote/httpmessage.hpp:
```diff
@@ -0,0 +1,297 @@
+			}
+
+			auto const wBuf = m_Body.m_Buffer.prepare(size);
+			boost::asio::buffer_copy(wBuf, buffers);
```
>Trivial? It's still not trivial :).

My definition of trivial is that the code doesn't do something exceptional, because it's a variation of the same thing you're already doing. All you need to add any number of similar things to `SendFile()` is use the prepare/commit logic of the dynamic buffer. The most complicated part here was getting std::ifstream to do what we want, not the dynamic buffer, and that could be simplified even more with `asio::stream_file` at some point in the future and with that would then be fully async, which is another advantage.

It's also about the same amount of (local) code you'd need to implement something with `file_body`, while not adding three different code paths for potential errors and platform issues. Add to that that you don't need to copy/move around of the header and it's definitely a cleaner and simpler solution than anything else I could think of.

jschmidt-icinga commented on test/base-tlsstream-fixture.hpp:
```diff
@@ -0,0 +1,107 @@
+				server->next_layer().async_handshake(handshake_type::server,
+					[&, this](const boost::system::error_code& ec) {
+						if (ec) {
+							BOOST_TEST_MESSAGE("Server Handshake Error: " + ec.message());
+							return;
```
Yes it probably would. It'll be a good idea to propagate an exception to the surrounding scope, so I'll do that here.

jschmidt-icinga commented on test/remote-httpserverconnection.cpp:
```diff
@@ -0,0 +1,487 @@
+// using namespace boost::unit_test;
+using namespace boost::unit_test_framework;
+
+struct HttpServerConnectionFixture : TlsStreamFixture,
+									 IcingaApplicationFixture,
+									 ConfigurationCacheDirFixture,
+									 TestLoggerFixture
```
I don't think this is actually wrong, but github displaying tabs as 8 spaces, unless I'm misinterpreting what you want changed here. Or do you mean that we shouldn't align with tabs at all but use spaces for alignment?

yhabteab replied:
Fine, I was just saying ;)!

jschmidt-icinga commented on test/remote-httpserverconnection.cpp:
```diff
@@ -0,0 +1,487 @@
+
+using namespace icinga;
+using namespace boost::beast;
+// using namespace boost::unit_test;
```
Just a leftover. But honestly, I appreciate your eye for detail. :+1:  My eyes seem to read past these things sometimes.

yhabteab commented on test/remote-sslcert-fixture.hpp:
```diff
@@ -0,0 +1,71 @@
+		}
+	}
+
+	auto EnsureCertFor(const std::string& name)
```
I just expect from a function named `Ensure*` to assert that something is true and returns a boolean or an exception if it doesn't. Maybe it's just me 🤷!

jschmidt-icinga commented:
- Added a few more asserts and made the tests use the parser-based overload of `http::read()` instead of the message one [^1]. 

[^1]: Because as @yhabteab found out, when it returns an error code, even when the message is successfully read and the error stems from reading an `end_of_stream` error into the read buffer, the message is never moved back out of the parser object. I have no idea why they built it that way, and I don't think it caused any errors or false negatives, but better being safe than sorry.

yhabteab reviewed (approved):
LFTM!

PS: I've updated the PR description to include before and after MEM usage graphs that demonstrate the difference in memory usage with this change. So, please have look at them :)!

julianbrost commented on test/base-testloggerfixture.hpp:
```diff
@@ -0,0 +1,94 @@
+	{
+		std::string pattern;
+		std::promise<bool> prom;
+		std::shared_future<bool> crutch;
```
Does this exist only for the case that the timeout expires and because in that case, the entry isn't removed from `m_Expects`?

julianbrost commented on test/remote-sslcert-fixture.cpp:
```diff
@@ -0,0 +1,27 @@
+
+BOOST_AUTO_TEST_SUITE(remote_certs)
+
+BOOST_FIXTURE_TEST_CASE(setup_certs, ConfigurationDataDirFixture)
+{
+	if (boost::filesystem::exists(CertificateFixture::m_PersistentCertsDir)) {
+		boost::filesystem::remove_all(CertificateFixture::m_PersistentCertsDir);
+	}
+}
```
`setup_certs` deletes all certificates? That sounds odd.

jschmidt-icinga replied:
It deletes any certificates or remainders of that directory that may exist before the "setup" is run. Essentially, it's a second cleanup run before the first test that uses this certificates. The certs then get created lazily when a test uses the actual Fixture. I couldn't just call one test for both setup and cleanup, because CMake doesn't support running a test twice, so I had to add two test cases that do the same thing.

If you have an issue with the name, I could rename it to something like `certs_fixture_prepare_directory` and `certs_fixture_cleanup_directory` or something like that. Additionally I could move the code to a static function so there is no duplication.

julianbrost commented on test/base-tlsstream-fixture.hpp:
```diff
@@ -0,0 +1,109 @@
+		tcp::acceptor acceptor{io, tcp::endpoint{address_v4::loopback(), 0}};
+		acceptor.listen();
+		acceptor.async_accept(server->lowest_layer(), [&](const boost::system::error_code& ec) {
+			if (ec) {
+				BOOST_TEST_MESSAGE("Server Accept Error: " + ec.message());
+				return;
+			}
```
This case ignores `p`.

julianbrost commented on test/base-tlsstream-fixture.hpp:
```diff
@@ -0,0 +1,109 @@
+		}
+
+		if (!client->next_layer().IsVerifyOK()) {
+			throw std::runtime_error{"Verfiy failed for connection."};
```
```suggestion
			throw std::runtime_error{"Verify failed for connection."};
```

julianbrost commented on test/base-tlsstream-fixture.hpp:
```diff
@@ -0,0 +1,109 @@
+		}
+
+		auto f = p.get_future();
+		f.wait();
+		f.get();
```
`get()` already waits, so no need to do this separately.

julianbrost commented on test/remote-sslcert-fixture.hpp:
```diff
@@ -0,0 +1,71 @@
+
+	auto EnsureCertFor(const std::string& name)
+	{
+		struct
+		{
+			String crtFile;
+			String keyFile;
+			String csrFile;
+		} cert;
```
I'd give this struct a proper name. Like this, my IDE only shows `<anonymous>` as the deduced type in the places where this method is used, which is a bit annoying.

jschmidt-icinga commented on test/base-testloggerfixture.hpp:
```diff
@@ -0,0 +1,94 @@
+
+struct TestLoggerFixture
+{
+	TestLoggerFixture()
+	{
+		testLogger->SetSeverity(testLogger->SeverityToString(LogDebug));
+		testLogger->Activate(true);
+		testLogger->SetActive(true);
+	}
```
Because there is nothing to clean up explicitly, so the auto-generated destructor should be enough.

julianbrost replied:
`Activate()` calls `Start()` on the object:

https://github.com/Icinga/icinga2/blob/4b1020e43c4c15893ed08a56acc24d8a090b7100/lib/base/configobject.cpp#L361-L368

And for loggers, `Start()` registers the logger:

https://github.com/Icinga/icinga2/blob/4b1020e43c4c15893ed08a56acc24d8a090b7100/lib/base/logger.cpp#L53-L60

So I don't see how this would be undone unless you explicitly stop/deactivate the object.

jschmidt-icinga replied:
Ok, I see. I couldn't find anywhere else this is done in the code, but that's because normally it'd be taken care of by `ConfigObject::StopObjects`, but that only gets called at the end of the main event loop. I'll add a destructor for now, but maybe in the future, when there are more Objects to be Stopped/Deactivated I'll see if I can add this functionality to the application fixture.

jschmidt-icinga commented on test/base-configuration-fixture.hpp:
```diff
@@ -0,0 +1,52 @@
+
+namespace icinga {
+
+struct ConfigurationDataDirFixture
+{
+	ConfigurationDataDirFixture()
+		: m_DataDir(boost::filesystem::current_path() / "data"), m_PrevDataDir(Configuration::DataDir.GetData())
+	{
+		Configuration::DataDir = m_DataDir.string();
+		boost::filesystem::create_directories(m_DataDir);
+	}
+
+	~ConfigurationDataDirFixture()
+	{
+		boost::filesystem::remove_all(m_DataDir);
+		Configuration::DataDir = m_PrevDataDir.string();
+	}
+
+	boost::filesystem::path m_DataDir;
+	boost::filesystem::path m_PrevDataDir;
+};
+
+struct ConfigurationCacheDirFixture
+{
+	ConfigurationCacheDirFixture()
+		: m_CacheDir(boost::filesystem::current_path() / "data"), m_PrevCacheDir(Configuration::CacheDir.GetData())
+	{
+		Configuration::CacheDir = m_CacheDir.string();
+		boost::filesystem::create_directories(m_CacheDir);
+	}
+
+	~ConfigurationCacheDirFixture()
+	{
+		boost::filesystem::remove_all(m_CacheDir);
+		Configuration::CacheDir = m_PrevCacheDir.string();
+	}
+
+	boost::filesystem::path m_CacheDir;
+	boost::filesystem::path m_PrevCacheDir;
+};
```
I didn't want to make things more complicated than they already are by adding getters/setters for everything. For now I'll look over it and move variables to private that aren't conceivably useful for writing tests, like `m_PrevCacheDir` here.

julianbrost commented on test/base-configuration-fixture.hpp:
```diff
@@ -0,0 +1,52 @@
+
+struct ConfigurationDataDirFixture
+{
+	ConfigurationDataDirFixture()
+		: m_DataDir(boost::filesystem::current_path() / "data"), m_PrevDataDir(Configuration::DataDir.GetData())
+	{
+		Configuration::DataDir = m_DataDir.string();
+		boost::filesystem::create_directories(m_DataDir);
+	}
+
+	~ConfigurationDataDirFixture()
+	{
+		boost::filesystem::remove_all(m_DataDir);
+		Configuration::DataDir = m_PrevDataDir.string();
+	}
```
I mean it's just swapping two lines in both functions. And configuring the directory only after it was created and deleting it only after it was unconfigured just sounds like the more logical order of operations here.

jschmidt-icinga replied:
My bad :facepalm:. I didn't read the current iteration (you quoted above) right. At some point I had a `std::exchange` in the constructor initializer list and thought I would have to add a temporary variable... You're correct, it makes more sense that way.

jschmidt-icinga commented on lib/remote/httpmessage.hpp:
```diff
@@ -0,0 +1,253 @@
+		template<typename T>
+		value_type& operator<<(T&& right)
+		{
+			/* Preferably, we would return an ostream object here instead. However
+			 * there seems to be a bug in boost::beast where if the ostream, or rather its
+			 * streambuf object is moved into the return value, the chunked encoding gets
+			 * mangled, leading to the client disconnecting.
```
Yes, definitely. I mentioned in the PR description that once I have some spare time I'll see if I can put together a minimal reproducer and report the issue.

jschmidt-icinga commented on lib/remote/httpmessage.cpp:
```diff
@@ -0,0 +1,166 @@
+
+void HttpResponse::SendFile(const String& path, boost::asio::yield_context yc)
+{
+	static constexpr std::size_t minChunkSize = 4096;
```
I agree, but I guess the logic was kind of to stay consistent with the naming of `m_MinPendingBufferSize` in the JSON output adapter, but thinking about it, it actually sense there, because it's the minimum size to cause a flush, whereas here "minimum size to be read" is plain false/misleading. Naming things is not my strength... :laughing:

jschmidt-icinga commented on lib/remote/httpmessage.cpp:
```diff
@@ -0,0 +1,166 @@
+	std::ifstream fp(path.CStr(), std::ifstream::in | std::ifstream::binary | std::ifstream::ate);
+	fp.exceptions(std::ifstream::badbit | std::ifstream::eofbit);
+
+	std::size_t remaining = fp.tellg();
```
The proper type would be `std::streamoff`, which `std::streampos` implicitly converts to and is a typedef of `std::uint64_t`, however since the veriable isn't really used as an offset, I agree that using `uint64_t` is probably the best choice.

jschmidt-icinga commented on lib/remote/httpserverconnection.hpp:
```diff
@@ -30,27 +31,27 @@ class HttpServerConnection final : public Object
 		const Shared<AsioTlsStream>::Ptr& stream, boost::asio::io_context& io);
 
-	void Disconnect(boost::asio::yield_context yc);
+	void Disconnect(boost::asio::yield_context yc, const String& reason = {});
```
`std::string_view` is always the first thing I consider. However I've used String here because it would need to be converted to that anyway, because we need the `operator+` overloads `string_view` doesn't have, leading to an additional `String` object to be created in cases where `String is actually passed to the function`.

Since this entire commit is up for discussion, I'll leave this as it is for now, same as the other comment about the reason message.

julianbrost commented on test/remote-certificate-fixture.hpp:
```diff
@@ -0,0 +1,71 @@
+/* Icinga 2 | (c) 2025 Icinga GmbH | GPLv2+ */
+
+#ifndef SSLCERT_FIXTURE_H
+#define SSLCERT_FIXTURE_H
```
And the include guard, which is now inconsistent with the file name :sweat_smile:

jschmidt-icinga replied:
Ok, time for `#pragma once`.

julianbrost commented on test/base-tlsstream-fixture.hpp:
```diff
@@ -0,0 +1,110 @@
+
+		tcp::acceptor acceptor{io, tcp::endpoint{address_v4::loopback(), 0}};
+		acceptor.listen();
+		acceptor.async_accept(server->lowest_layer(), [&](const boost::system::error_code& ec) {
+			if (ec) {
+				BOOST_TEST_MESSAGE("Server Accept Error: " + ec.message());
+				return;
+			}
+			server->next_layer().async_handshake(handshake_type::server, [&](const boost::system::error_code& ec) {
+				if (ec) {
+					BOOST_TEST_MESSAGE("Server Handshake Error: " + ec.message());
+					p.set_exception(std::make_exception_ptr(boost::system::system_error{ec}));
+					return;
+				}
+
+				p.set_value();
+			});
+		});
```
If you check `IsVerifyOK()` on the client side, you probably want to do it here as well given that you supply a client certificate. (And probably `GetPeerCertificate() != nullptr` as well.)

julianbrost replied:
Ah right, wasn't sure what `IsVerifyOK()` would do if no client certificate was presented. But apparently it already does the safer thing.

jschmidt-icinga commented:
+ Replaced the `DetectClientShutdown()` coroutine with a renamed version of the previous `StartStreaming()` one, that is again invoked through the handler. We've experimented with different ways of handling shutdowns, but ultimately the decision was to stick to what worked before. I'm not happy with this, I think there could have been a better, more general solution, but that arguably would have been more complex too.
+ Addressed most of the other comments, except (unless I'm missing something else) the Param test case and reusing the message object in SendJsonError (which I'm still unsure about). I'll work on those tomorrow, but I wanted to push today to see if all the checks complete successfully.

jschmidt-icinga commented on test/remote-httpmessage.cpp:
```diff
@@ -0,0 +1,279 @@
+
+BOOST_AUTO_TEST_CASE(request_params)
+{
+	HttpRequest request(client);
+	request.body() = JsonEncode(new Dictionary{{"test1", false}, {"test2", true}});
+	request.target("https://localhost:1234/v1/test?test1=1&test3=0&test3=1&test4=0");
+	request.DecodeParams();
+
+	auto params = request.Params();
+	BOOST_REQUIRE(params);
+	BOOST_REQUIRE(!params->Get("test2").IsObjectType<Array>());
+	BOOST_REQUIRE(params->Get("test2").IsBoolean());
+	BOOST_REQUIRE(params->Get("test2"));
+	BOOST_REQUIRE(params->Get("test1").IsObjectType<Array>());
+	BOOST_REQUIRE(HttpUtility::GetLastParameter(params, "test1"));
+	BOOST_REQUIRE(params->Get("test3").IsObjectType<Array>());
+	BOOST_REQUIRE(HttpUtility::GetLastParameter(params, "test3"));
+	BOOST_REQUIRE(params->Get("test4").IsObjectType<Array>());
+	BOOST_REQUIRE(!HttpUtility::GetLastParameter(params, "test4"));
```
Testing with booleans is actually necessary to ensure overriding booleans with "0" and "1" in the URL works.

However I will add string values to test the precedence, rename everything to make the intent more clear and add comments on what is actually being tested, because I agree that its far from self-explanatory.

julianbrost reviewed (commented):
That's my review for today (didn't go through all the new changes to the tests so far, but only where I added comments previously)

jschmidt-icinga commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -99,11 +98,11 @@ void HttpServerConnection::Disconnect(boost::asio::yield_context yc)
 
-void HttpServerConnection::StartStreaming()
+void HttpServerConnection::StartDetectClientSideShutdown()
 {
 	namespace asio = boost::asio;
 
-	m_HasStartedStreaming = true;
+	m_ConnectionReusable = false;
```
Will this do: `// all hope abandon ye who enter here` :laughing:

jschmidt-icinga commented on test/remote-httpmessage.cpp:
```diff
@@ -0,0 +1,328 @@
+BOOST_AUTO_TEST_CASE(request_params)
+{
+	HttpRequest request(client);
+	request.body() = JsonEncode(new Dictionary{{"bool-in-json", true}, {"bool-in-url-and-json", true},
+		{"string-in-json", "json-value"}, {"string-in-url-and-json", "json-value"}});
+	request.target("https://localhost:1234/v1/test?"
+				   "bool-in-url-and-json=0&"
+				   "bool-in-url=1&"
+				   "string-in-url-and-json=url-value&"
+				   "string-only-in-url=url-value");
```
>Are we in fighting clang-tidy territory here? :D

You mean clang-format, but yes. I manually formatted it one per line, but I now do a pass over all files newly added by this PR before committing and it must have reformatted it that way. What works is adding `//` at the end of each manual line break or turning it on for a block with `// clang-format on/off`, but those are ugly too. :shrug:

jschmidt-icinga commented on test/remote-httpserverconnection.cpp:
```diff
@@ -0,0 +1,555 @@
+	request.set(http::field::accept, "application/json");
+	request.set(http::field::connection, "close");
+	request.content_length(0);
+	http::write(*client, request);
+	client->flush();
+
+	flat_buffer buf;
+	http::response<http::string_body> response;
+	BOOST_REQUIRE_NO_THROW(http::read(*client, buf, response));
+
+	BOOST_REQUIRE_EQUAL(response.version(), 11);
+	BOOST_REQUIRE_EQUAL(response.result(), http::status::ok);
+
+	BOOST_REQUIRE(Shutdown(client));
+}
+
+BOOST_AUTO_TEST_CASE(error_authenticate_passwd)
+{
+	CreateTestUsers();
+	SetupHttpServerConnection(false);
+
+	http::request<boost::beast::http::string_body> request;
+	request.method(http::verb::get);
+	request.target("https://localhost:5665/v1/test");
+	request.set(http::field::authorization, "Basic " + Base64::Encode("test:test"));
+	request.set(http::field::accept, "application/json");
+	request.set(http::field::connection, "close");
+	request.content_length(0);
+	http::write(*client, request);
+	client->flush();
+
+	flat_buffer buf;
+	http::response<http::string_body> response;
+	BOOST_REQUIRE_NO_THROW(http::read(*client, buf, response));
+
+	BOOST_REQUIRE_EQUAL(response.version(), 11);
+	BOOST_REQUIRE_EQUAL(response.result(), http::status::ok);
+
+	BOOST_REQUIRE(Shutdown(client));
+}
```
Right, I'll switch it around to be `authenticate_` and then `authenticate_error` for the ones that actually test for error.

jschmidt-icinga commented on test/remote-httpserverconnection.cpp:
```diff
@@ -0,0 +1,555 @@
+	BOOST_REQUIRE_EQUAL(parser.get().result(), http::status::ok);
+	BOOST_REQUIRE_EQUAL(parser.get().body(), "test");
+
+	if (!ec) {
+		/* Sometimes the call to read above already catches the end_of_stream error, sometimes
+		 * it stops just shy of that byte and we'll want to do another dummy read to catch it
+		 * here.
+		 */
+		http::response<http::string_body> response{};
+		http::read(*client, buf, response, ec);
+	}
```
My reasoning is that with `reuse_connection`, the thing we actually test is that the connection is *not* closed after the first message and then a second response is received. Afterwards we don't care as much about who initiates the shutdown, even though it will likely be the server side.

Whereas with `wg_abort`, the focus of the testing is not on the message, but that the server initiates the shutdown, which means, we ensure that the client receives a close_notify, then we assert that `HttpServerConnection::IsDisconnected()` is true, and then we complete the shutdown successfully on our side as a formality.

That said, I can add the same thing for the `reuse_connection` too, it shouldn't hurt anything.

julianbrost replied:
My thinking was that the connection shutdown between these two test cases looks very similar but differs in this detail. Given that the comment says "sometimes", this could mean that this is a very low probability failure mode and it was only added here because you've only observed it here.

Though now I've tried locally what happens if I remove this `if` block. Result: the test fails every single time, so doesn't look low-probability. Which got me thinking: should it even happen that this if is skipped with `ec == end_of_stream` from the previous `http::read()`? That was the last response and reading it should always succeed, shouldn't it?

If there was a scenario where it would fail, couldn't the final `http::read()` in `reuse_connection` throw `end_of_stream` as an exception (there the function is called without an `ec` parameter)?

So either the test should also work reliable with a `BOOST_REQUIRE(!ec);` before this if (and thus the read currently in the `if` is done unconditionally), or I'm still understanding something wrong.

julianbrost replied:
While playing around with this, I noticed another flaw. Not sure if inside the test or inside the implementation. I think I'm also missing a clear statement what this test case is actually supposed to test. Is it that after the wait group is joined, the connection is closed or that the current response is still finished?

Anyway, the behavior I notices is that if you add a `Utility::Sleep(1);` right after `client->flush();` (so that by the time the test continues, the server should have finished processing the response and all of it is already in our (kernel) receive queue), the test case takes 15 seconds and the connection is terminated due to inactivity:

```
[2025-08-22 14:25:37 +0000] information/HttpServerConnection: Request GET https://localhost:5665/v1/test (from [127.0.0.1]:47820, user: client, agent: , status: OK) took total 0ms.
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(411): info: check 'no exceptions thrown by http::read(*client, buf, parser)' has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(413): info: check parser.is_header_done() has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(414): info: check parser.is_done() has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(415): info: check parser.get().version() == 11 has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(416): info: check parser.get().result() == http::status::continue_ has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(419): info: check buf.size() == 0 has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(425): info: check !ec has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(427): info: check parser.is_header_done() has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(428): info: check parser.is_done() has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(429): info: check parser.get().version() == 11 has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(430): info: check parser.get().result() == http::status::ok has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(431): info: check parser.get().body() == "test" has passed

[2025-08-22 14:25:52 +0000] information/HttpServerConnection: No messages for HTTP connection have been received in the last 10 seconds.
[2025-08-22 14:25:52 +0000] information/HttpServerConnection: HTTP client disconnected (from [127.0.0.1]:47820)
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(441): info: check ec == boost::system::error_code{boost::beast::http::error::end_of_stream} has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(443): info: check AssertServerDisconnected(std::chrono::seconds(5)) has passed
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(444): info: check Shutdown(client) has passed. Error: Success
/home/jbrost/src/icinga2/test/remote-httpserverconnection.cpp(445): info: check ExpectLogPattern("HTTP client disconnected .*", std::chrono::seconds(5)) has passed
```

jschmidt-icinga replied:
>IMHO that would be a highly surprising interface

Yes, we're using boost::beast.

>This isn't backed by the [documentation for `http::read()`](https://www.boost.org/doc/libs/1_66_0/libs/beast/doc/html/beast/ref/boost__beast__http__read/overload2.html):

This might come down to the fine print, where nothing is parsed yet **in this invocation of `http::read_some`**, even though bytes have been read from the stream that are enough to finish parsing the message, the error will be returned and the bytes will not be given to the parser. I don't think this is documented anywhere, but you can read the implementation of `http::read_some()` (which `http::read` also uses).

If you do something like the following, you can see that the parser will not complete successfully.

``` C++
	while(!parser.is_done() && !ec){
		Utility::Sleep(0.5);
		http::read_some(*client, buf, parser, ec);
		// BOOST_REQUIRE_EQUAL(ec, boost::system::errc::success);
	}
```

Granted though, this makes the test-case still defective on this side of the race condition, because the following asserts will fail. I'll look into how I can fix that.

>Is it that after the wait group is joined, the connection is closed or that the current response is still finished?

Both. Since we send a request with `keep_alive`, we want to know if the server goes against that and terminates the connection after sending the message. Practically the opposite of `reuse_connection`.

>Anyway, the behavior I noticed is that if you add a `Utility::Sleep(1);` right after `client->flush();` (so that by the time the test continues, the server should have finished processing the response and all of it is already in our (kernel) receive queue),

Right. That's what we talked about two weeks ago which is why I added the 100-continue part (which you actually suggested :laughing: ), but obviously that just delays the race condition we had. This is probably the most difficult test case in this suite. I will look into how to fix it properly on Monday.

jschmidt-icinga replied:
That line is only in the documentation from boost-1.66 to 1.69. I don't know if it was removed because its a documentation error, or because the behavior of the code changed and removed this requirement, but I'm going to assume the former. If you read the current source, calling consume on that buffer in between `read_some()` calls makes no sense because the consume is done internally. If anything, it would break reading subsequent messages.

However, it seems the error `http:7` is actually `buffer_overflow`, not as I assumed `end_of_stream`, and goes away when i increase max_size(), which I don't understand. `read_some()` should work with with any buffer size as long as there is more than 0 space left. In any case, I can't reproduce the `end_of_stream` error on reading that message, so I will make the second read unconditional as you suggested.

jschmidt-icinga commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -99,11 +98,38 @@ void HttpServerConnection::Disconnect(boost::asio::yield_context yc)
+ * @remarks @parblock
+ * Technically it would be possible to detect disconnects on the TCP-side by setting the
+ * socket to non-blocking and then performing a read directly on the socket with the message_peek
+ * flag. As the TCP FIN message can only arrive when the queue isn't full, this would technically
+ * be safe and free of side-effects.
```
>Isn't that a reason for actually reading (as we did and continue to do) over peeking?

My mistake, I wanted to say exactly the opposite there, but then somehow got it turned around in my head. 

I think FIN arrives and changes the connection state to CLOSE_WAIT, regardless of how many bytes remain in the buffer. Which is also what I meant by safe, we can check if the connection has been closed, even if the kernel buffer contains an arbitrary number of bytes. Unlike the TLS-side and its user-space buffer, which we can only read into until its full, with a possible close_notify remaining in the kernel buffer, inaccessible to us.

jschmidt-icinga commented:
+ Changed the `wg_abort` test case to use a new "wait" parameter to the `UnitTestHandler` that lets the test case wait until the handler is called by `HttpServerConnection`. This way we can make sure there are no race-conditiions.
+ Fixed the inconsistencies between the different `http::read` calls. Now BOOST_REQUIRE_NO_THROW is used everywhere where we don't care about the error code and explicitly check it.
+ Moved the remark comment inside `StartDetectClientSideShutdown` and corrected what i'd written about FIN.

jschmidt-icinga commented:
+ Changed the handler as suggested to register lambdas callable via url path.
+ Adjusted `target()` and added host field as suggested.

julianbrost commented on test/remote-httpserverconnection.cpp:
```diff
@@ -0,0 +1,557 @@
+	http::response_parser<http::string_body> parser;
+	BOOST_REQUIRE_NO_THROW(http::read(*client, buf, parser));
+
+	BOOST_REQUIRE(parser.is_header_done());
+	BOOST_REQUIRE(parser.is_done());
+	BOOST_REQUIRE_EQUAL(parser.get().version(), 11);
+	BOOST_REQUIRE_EQUAL(parser.get().result(), http::status::ok);
```
I'd also test that a body is also passed correctly. This should make the test more robust in the sense that the version and status are default values, so unless they are set to something else explicitly, these checks will always pass.

julianbrost commented on test/remote-httpserverconnection.cpp:
```diff
@@ -0,0 +1,569 @@
+	http::write(*client, sr);
+	client->flush();
+
+	http::read(*client, buf, response);
```
> Is there a reason why the one inside `response_write_empty` didn't get the `BOOST_REQUIRE_NO_THROW` treatment?

Seems like you might have overlooked this comment. That one call still looks out of line when I do a quick `git grep http::read`:

https://github.com/Icinga/icinga2/blob/0d9526c6d61884c7f0e73cd081b576affbb87958/test/remote-httpmessage.cpp#L179

jschmidt-icinga replied:
Yes, I missed this one because I was only looking at `remote_httpserverconnection.cpp`, despite you clearly mentioning the exact test case. Sorry.

jschmidt-icinga replied:
Actually that one is the correct one and somewhere in my mindless "addressing" I must have overlooked why I had left it out here in the first place. An assert here would potentially be concurrent with the ones in the coroutine, which Boost.Test does not support. This is quite unintuitive and more modern frameworks like doctest solve this by making asserts thread-safe, but on boost we have to ensure no asserts can be running concurrently. And here (before I messed it up again) I'm using Shutdown as a synchronization point between where asserts are done inside and outside the coroutine.

So instead of adding `BOOST_REQUIRE_NO_THROW` here, I will be removing it everywhere else and using the error_code overload of `http::read()` and an assert on that error code after shutdown instead. Technically, letting the function throw would also be ok, since it would terminate the test with a clear message, but I think it's better to add asserts on the error_code.

julianbrost commented on lib/remote/httpmessage.cpp:
```diff
@@ -0,0 +1,193 @@
+{
+}
+
+void HttpResponse::Clear()
+{
+	ASSERT(!m_SerializationStarted);
+	result(boost::beast::http::status::ok);
+	version(11);
+	base().clear();
+	body() = body_type::value_type{};
+}
```
Just had another thought about this: do we really have to list everything that needs clearing manually here? `http::response` provides an `operator=`, so would there be anything wrong with just using this (similar to how your implementation already does for the body)?

```suggestion
void HttpResponse::Clear()
{
	ASSERT(!m_SerializationStarted);
	boost::beast::http::response<body_type>::operator=({});
}
```

Passes the unit tests on my machine.

julianbrost commented:
I've spun this up for a test (see below for the details) and while I could confirm significant memory savings (-87%), I also noticed quite an increase in response times (+132%) and CPU usage (+90%) when using `/v1/objects`. Take the exact numbers with a grain of salt, they are just from a single run, but other runs showed them in similar orders of magnitude. I expected some tradeoff for the reduced memory usage (with the chunk headers, there's a bit more data transfer, we're making more round-trips through the HTTP implementation), but that's quite a bit more than what I'd have hoped for. Can you reproduce this? Is this to be expected?

## Test setup

I've written a short shell script that can be simply mounted into our container images and automatically executes a test for it which creates a config with an ApiListener and 10k Host objects (and disables the checker and notification features to take them out of the equation). It then performs a few (4) requests to /v1/objects/hosts. The daemon is started wrapped in GNU time which (in contrast to the shell built-ins) also shows peak memory usage (this needs some extra fiddling around since this isn't installed by default).

```sh
#!/usr/bin/env bash

set -euo pipefail

apt update
apt install -yy time
chown -R icinga:icinga /data/

icinga2 api setup
icinga2 feature disable checker
icinga2 feature disable notification
pw=$(grep password /etc/icinga2/conf.d/api-users.conf | cut -d\" -f2)
api_req() {
	curl -Ssku "root:$pw" "$@"
}

cat > /etc/icinga2/conf.d/hosts.conf <<EOF
for (var i in range(10000)) {
	object Host i {
		check_command = "dummy"
	}
}
EOF

env time -v icinga2 daemon &
daemon_pid=$!
sleep 2 # wait for startup


curl_pids=()
for i in {1..4}; do
	api_req 'https://localhost:5665/v1/objects/hosts' -X GET --json '{"pretty":true}' -o /dev/null &
	curl_pids+=($!)
done

wait "${curl_pids[@]}"

api_req 'https://localhost:5665/v1/actions/shutdown-process' --json '{"pretty":true}'

wait "$daemon_pid"
```

It can be run like this:

```sh
docker run --rm -it -u root \
    --mount type=bind,src=$(pwd)/test-objects-query-mem.sh,dst=/test.sh \
    icinga/icinga2:v2.15.0-68-g25b03b2ff \
    /test.sh
```

Below are excerpts of the relevant parts of the output.

Note: I did the tests based on the PR before the last push to it, but that changed only something inside the tests so it doesn't matter for my tests.

### PR base (1f92ec656bca6ffcbda6638b043946d437084e7b)

```
Request GET /v1/objects/hosts ([...], status: OK) took total 2933ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 3014ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 3046ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 3057ms.
```

```
	User time (seconds): 16.63
	System time (seconds): 1.15
```

```
	Maximum resident set size (kbytes): 709052
```

### PR (25b03b2ffc27c330c5066d42d142a8834fe3f2d0)

```
Request GET /v1/objects/hosts ([...], status: OK) took total 6979ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 6988ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 6996ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 7003ms.
```

```
	User time (seconds): 33.28
	System time (seconds): 0.59
```

```
	Maximum resident set size (kbytes): 88572
```

jschmidt-icinga commented:
I wouldn't have expected a 90% CPU-time increase, but a significant increase still. Potentially this could be significantly lowered by increasing `HttpResponseJsonWriter::m_MinPendingBufferSize` to a value that still is negligible in memory consumption but needs a fraction of the round-trips.

I'll try to get your script running with my setup and see if that can help.

jschmidt-icinga commented:
Ok, so from my preliminary testing, I can only reproduce a difference of this magnitude without optimization. Enabling `-O2 -DNDEBUG` severely cuts down on the impact on CPU-time, while increasing the flush-threshold to around 64k is reducing it to ~0 (I actually found a slight improvement in user time with even larger thresholds) on my system. Response time seems to mostly be unchanged between 1f92ec656 and this PR when optimizations are enabled regardless of flush threshold.

Maybe you can repeat the test with optimization on, on your end and see if it's still a problem. I'll do a few more tests tomorrow and find the ideal value for the flush threshold.

yhabteab commented:
I have also tested this as described in https://github.com/Icinga/icinga2/pull/10516#issuecomment-3228271848 and got almost the same results!
```bash
Request GET /v1/objects/hosts (from [::1]:38850, user: root, agent: curl/8.14.1, status: OK) took total 5076ms.
Request GET /v1/objects/hosts (from [::1]:38818, user: root, agent: curl/8.14.1, status: OK) took total 5113ms.
Request GET /v1/objects/hosts (from [::1]:38840, user: root, agent: curl/8.14.1, status: OK) took total 5185ms
```

And profiling it with dtrace on my Mac shows that the `write_character(s)` methods of the output adapter take a lot of CPU than I would expect them to, since this is basically transferring data from one to another buffer. However, the multi_buffer::prepare and buffer_copy functions seem to do a lot of work each time.

<img width="1762" height="840" alt="Bildschirmfoto 2025-08-27 um 17 45 09" src="https://github.com/user-attachments/assets/2b6e425c-fcd4-4f17-b1a2-9421016eba6a" />
<img width="1596" height="147" alt="Bildschirmfoto 2025-08-27 um 17 53 41" src="https://github.com/user-attachments/assets/cfb1a4d6-486b-468b-94d1-a3743aff8f32" />

<img width="1647" height="721" alt="Bildschirmfoto 2025-08-27 um 17 54 11" src="https://github.com/user-attachments/assets/43f50c3d-33d8-412e-abac-3c5d4d3f2828" />

jschmidt-icinga commented:
>Asio's streambuf doesn't have such constraints Calling members of the underlying buffer before the output stream is destroyed results in undefined behavior., so it uses a persistent std::ostream member instead of creating a new one for each << operation.

That isn't relevant here, since everything is going through the output adapter and not the ostream operator. My guess (since from your numbers I'm assuming you're still running the test unoptimized) would be that some low level functionality of asio's streambuf is compiled into the dynamic libraries (with optimization) and that's why its faster. As I said before, `streambuf` uses multi_buffer internally and I'd be surprised if it was any faster **with optimizations enabled**. 

This all comes down on whether it's a priority for us that our binary should not have performance regressions in an unoptimized debug binary (I didn't test `-Og`, but I'd bet it's still a lot closer than your results).

I'll finish a few more tests and then I'll get back to you with my own results.

yhabteab commented:
> This all comes down on whether it's a priority for us that our binary should not have performance regressions in an unoptimized debug binary (I didn't test `-Og`, but I'd bet it's still a lot closer than your results).

I didn't use debug binary for my previous tests! I just built an image with #10505 and run it as described in https://github.com/Icinga/icinga2/pull/10516#issuecomment-3228271848. That's the exact executable that will also be used by the end users if they use our image.

> As I said before, `streambuf` uses multi_buffer internally and I'd be surprised if it was any faster **with optimizations enabled**.

Are sure about this? Beast dependes on Asio's functionality but never the other way around. Asio's streambuf implements the `std::streambuf` interface, so it has its own techniques to manage the buffers and doesn't depend on any Beast's functionality.

jschmidt-icinga commented:
>Beast dependes on Asio's functionality but never the other way around. Asio's streambuf implements the std::streambuf interface, so it has its own techniques to manage the buffers and doesn't depend on any Beast's functionality.

My bad. I must have misremembered. I thought this was part of beast. Seems it uses a `std::vector` inside.

>I didn't use debug binary for my previous tests!

Then why do you get these absurdly high response times? Mine are more along the lines of 610ms per response for 1f92ec656 (on my potato notebook mind you) and my PR starting at ~690ms at a 4k flush threshold (and decreasing significantly on higher thresholds).

julianbrost commented:
@jschmidt-icinga You're right, compiler optimizations (or the lack thereof) is to blame here. Only adding `-O2` to `CXXFLAGS` changed the numbers for me as follows:

### Base (1f92ec656bca6ffcbda6638b043946d437084e7b)

```
Request GET /v1/objects/hosts ([...], status: OK) took total 876ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 885ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 887ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 889ms.
```

```
	User time (seconds): 4.08
	System time (seconds): 0.68
```

```
	Maximum resident set size (kbytes): 699536
```

### PR (bf04a5588172d41e6c06a1f8fefb2047dfa1da1f)

```
Request GET /v1/objects/hosts ([...], status: OK) took total 888ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 889ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 893ms.
Request GET /v1/objects/hosts ([...], status: OK) took total 893ms.
```

```
	User time (seconds): 4.58
	System time (seconds): 0.42
```

```
	Maximum resident set size (kbytes): 81212
```

So with that, it's just a 0.7% increase in response time and 5% increase in CPU usage for a 88% decrease in memory usage (again, take the exact numbers with a grain of salt, they are from a single run, repeating it shows similar numbers though). That sounds more than acceptable for that saving in memory.

> This all comes down on whether it's a priority for us that our binary should not have performance regressions in an unoptimized debug binary

So that in itself wouldn't be a problem for me.

Though I haven't looked in detail into what @yhabteab found, should we have a closer look at this nonetheless?

jschmidt-icinga commented:
I've got the following results with a container compiled with `-O2 -DNDEBUG`. Per test I ran the test script three times, averaged the four response times per run and then all values over the three runs. For each run I've doubled the value of `HttpResponseJsonWriter::m_MinPendingBufferSize`, from 4k to 128k, and 1M as an additional test.


| test                  | response time (ms) | user time |  mem |
|--------------------- | ------------------ | --------- | ------- |
| 1f92ec656             |               611 |      2.96 | 707074 |
| 4k                    |               690 |      3.54 |  78473 |
| 8k                    |               651 |      3.47 |  79154 |
| 16k                   |               596 |      3.26 |  78977 |
| 32k                   |               563 |      3.10 |  79718 |
| 64k                   |               543 |      3.01 |  80483 |
| 128k                  |               534 |      2.98 |  78865 |
| 1M                   |               532 |      2.94 |  86795 |

At 1M, which I initially did just so make sure a really big number doesn't get any additional gains, I did an additional test that only does `async_flush()` conditionally when the serializer is done, and that looked promising, so I ran it again on lower threshold values, where I would have expected flushing manually to matter more, but it didn't seem to do anything, so I left those results out here.

I also want to note that the values for this PR were incredibly consistent compared to 1f92ec656 (likely because of fewer memory allocations) which fluctuated a lot more so I averaged over 6 tests for that instead of 3 for the others.

As you can see, the response time actually gets better than 1f92ec656 while user time breaks even at about 64k/128k. Given that under 1M max memory usage doesn't seem to be impacted much, I'd actually set this (and the threshold in `SendFile()`) to 128k unless anyone has objections.

julianbrost commented:
> I'd actually set this (and the threshold in `SendFile()`) to 128k unless anyone has objections.

Sounds fine for me. (Afterwards I'll spin it up for a test again, but I wouldn't expect any big surprises there).

> The interface is very similar to the DynamicBuffer interface (which is the reason it worked here as an almost-drop-in replacement) and while the buffer it uses internally is contiguous, as opposed to `multi_buffer`, which is more `std::deque`-like

Fun fact: I'm reading the documentation of [`asio::streambuf`](https://www.boost.org/doc/libs/1_89_0/doc/html/boost_asio/reference/streambuf.html) as "we currently do that but if we feel like it, we could change it to something more like `multi_buffer` at any time":

> The basic_streambuf class's public interface is intended to permit the following implementation strategies:
>
> - A single contiguous character array, which is reallocated as necessary to accommodate changes in the size of the character sequence. This is the implementation approach currently used in Asio.
> - A sequence of one or more character arrays, where each array is of the same size. Additional character array objects are appended to the sequence to accommodate changes in the size of the character sequence.
> - A sequence of one or more character arrays of varying sizes. Additional character array objects are appended to the sequence to accommodate changes in the size of the character sequence.

jschmidt-icinga commented:
>The new Containerfile from #10505 sets the `CMAKE_BUILD_TYPE` to [`ReleaseWithDebInfo`](https://github.com/Icinga/icinga2/pull/10505/files#diff-5fcdf9b4580789697d834d1456a22bcfaa236d668fc180cad4775afc36ed5914R91) but that's not a valid CMake flag

I find this funny because in my last comment, I first wrote `RelWithDebugInfo` instead before I edited it. :laughing: 

> However, in debug builds (which I mistakenly assumed it's a release build xD), Asio's implementation appears to be superior to Beast's implementation.

I think I found the reason why. look for `debug_check()` in the `multi_buffer` source. This puts asserts on basically all variables of the buffer at multiple points during each `consume()`/`commit()`/`prepare()` operation. `boost::asio::streambuf` on the other hand implements these functions using the the `setg()`/`gptr()`/`pptr()` interface of the `std::streambuf` base, which have no asserts at all.

julianbrost commented:
> I think I found the reason why. look for `debug_check()` in the `multi_buffer` source.

I doubt that's the culprit. They are all inside `#if BOOST_BEAST_MULTI_BUFFER_DEBUG_CHECK` and looks like you'd have to set that explicitly, I didn't find any place in the Boost headers where this was set (and neither do we set it inside Icinga 2).

julianbrost reviewed (approved):
> Afterwards I'll spin it up for a test again, but I wouldn't expect any big surprises there

Just did that, unsurprisingly it still worked fine, so for me, this PR is ready :rocket: 

Note: that doesn't mean that the code added by this PR has to be immutable, there may still be room for improvement, but I think the current state is fine and any further changes would be easier to grasp in follow-up PRs rather than inside a 2k+ lines and 200+ comments PR.
